### ISSUE TITLE
MTAG N (max) and other sample size estimates differ from expectation

### ISSUE THREAD

**Issue by carbocation** [omeed-maghzian/mtag#55]
I have 3 GWASes on related traits in tens of thousands of people. After MTAG completes, I see either tiny or huge values for N (max):

```
Summary of MTAG results:
------------------------
  Trait                 N (max)  N (mean)         ...           GWAS mean chi^2  MTAG mean chi^2  GWAS equiv. (max) N
1  trait1.gz-mtag.gz       29       25          ...           1.063            1.063                 29            
2  trait2.gz-mtag.gz       83       71          ...           1.066            1.068                 86            
3  trait3.gz-mtag.gz  5725973  4883237          ...           1.062            1.063            5774601            
```

The mean Chi^2 for the 3 traits are all around ~1.06 so I do get: `Mean chi^2 of SNPs used to estimate Omega is low for some SNPsMTAG may not perform well in this situation.` But I think I'm still surprised that:
1. The N (max) for two of the traits is listed as a value that is ~100-1000x smaller than the real N, and
1. The N (max) for the third trait is over 100x larger than the real N

Any pointers for troubleshooting? 

[Log is attached](https://github.com/omeed-maghzian/mtag/files/2609746/mtag.public.log)

**Comment by carbocation** [comment]
I've done another 10 GWASes on various traits in this same subpopulation to try to figure out why this extreme N displacement is happening here, and there are some standout features that might be relevant:

1. I am not using standardized phenotypes in my GWAS. Instead, the phenotypes keep their real values.
1. The trait that is gaining all of the N is one that is naturally in the same range as if it were a standardized phenotype. The rest of the traits are real-valued numbers that are in the 10s-100s.

This could just be a coincidence, but it persists across various trait pairings, so I'm a bit suspicious that it's relevant.

**Comment by carbocation** [comment]
Scaling up the small phenotype to be in the same range as the others:
```
Summary of MTAG results:
------------------------
  Trait                 N (max)  N (mean)  # SNPs used  GWAS mean chi^2  MTAG mean chi^2  GWAS equiv. (max) N
1  trait1.gz-mtag.gz   29       25       7121207      1.062            1.075             35                
2  trait2.gz-mtag.gz   83       71       7121207      1.064            1.072             92                
3  trait3.gz-mtag.gz  572      488       7121207      1.062            1.063            585                
```

So, maybe I don't actually understand what `N` means here.

**Comment by paturley** [comment]
This is weird. The N (max) column should represent the maximum value of the sample size column from your data. Are you sure that your data is reading the correct column or that the N column in your input files clean. Is it possible that the table got messed up in the latest update, @huilisabrina?

**Comment by carbocation** [comment]
Patrick, thanks for taking a look. Header & example SNP from an input file:
```
SNP	CHR	BP	ALLELE1	ALLELE0	A1FREQ	INFO	BETA	SE	P	N
rs367896724	1	10177	A	AC	0.599895	0.467935	0.315301	0.416004	3.8E-01	16894
```

Numerically sorting the N column and looking at the top and bottom values:
```
zcat trait1.gz-mtag.gz | cut -f11 | sort -g | head
N
16894
16894
16894
16894
16894
16894
16894
16894
16894
```

```
zcat trait1.gz-mtag.gz | cut -f11 | sort -g | tail
16894
16894
16894
16894
16894
16894
16894
16894
16894
16894
```
So at least what I can say is that:
1. All of the values of N are exactly 16894.
1. It seems that I'm getting the right column, since builtin unix tools are showing me what I expect for the "N" column.

**Comment by carbocation** [comment]
Also, I'm running mtag at commit a95fcc2cd1920543302e375ea27b36719848cc80 which is HEAD as far as I'm aware.

**Comment by paturley** [comment]
This is very curious. @huilisabrina may have to dig into this when she gets back into the office on Monday. I don't think I've seen this show up before, so maybe something got mixed up in the latest commit.

**Comment by huilisabrina** [comment]
Hi @carbocation ,

Thanks for your feedback on this. So the main reason here is that because you're specifying the flag `--use-beta-se`, the software turns to calculate the N column using your  _**SE and A1FREQ**_ columns provided, and then use this _**calculated**_ N as input to the MTAG estimator to weigh the Z scores. The summary table reports the max & mean of this converted N as opposed to the original N. 

Your two observations about phenotype standardization is indeed related! However this is an upshot of the conversion issue described above, not the cause of the problem. The more unstandardzized the phenotypes are, the more difference there will be between this converted N and the original N. But this is _**not**_ saying that you need to standardize the phenotypes before `mtag`.  Using `--use-beta-se` actually ensures the unit of beta/se between input and output to be the same. 

To verify the above, you can try adding a Z column to the GWAS input (using the existing beta and se). Then remove `--ues-beta-se`, `--beta-name` and `--se-name` in your `mtag` command. The output log should give you the correct mean/max of N. (However, without `--use-beta-se`, mtag will give results corresponding to if the phenotype had been standardized.) 

This is a very helpful feedback, because maybe I should change the table to report the mean / max of the _**raw**_ sample size, as opposed to the inferred one. I can also add another flag to make this more flexible / less confusing. @paturley let me know what you think about this too.

The bottom line is that the current sumstats should be correct - the N column in the output sumstats is the same as the input, the BETA/SE in the output are on the same unit as the input. I will fix the oddity in the summary table. By the way, I have seen you other comments and will get to them soon as well! 

Thanks,
Hui

**Comment by huilisabrina** [comment]
Hi @carbocation,

Thanks for testing and glad it was helpful! Sorry for being misleading about this. This should be fixed now in the latest commit 27639cf6fca047ca6fa2f35d99945534336865df . Could you try repulling and running `mtag` with `--use_beta_se`? This time I believe both the sumstats and the log table should give reasonable N info. Feel free to ask more questions! 

Best,
Hui

**Comment by carbocation** [comment]
Dear Hui,

Thanks! I tried to re-run, but got an error because `--drop_ambig_snps` is no longer an option. Removed that and re-ran. Now crashing with `ValueError: too many values to unpack`:

```
Dropped 1352017 SNPs due to strand ambiguity, 7286239 SNPs remain in intersection after merging trait1
Dropped 0 SNPs due to strand ambiguity, 7286239 SNPs remain in intersection after merging trait2
Dropped 0 SNPs due to strand ambiguity, 7286239 SNPs remain in intersection after merging trait3
Dropped 15 SNPs due to inconsistent allele pairs from phenotype 4. 7283689 SNPs remain.
Dropped 0 SNPs due to strand ambiguity, 7283689 SNPs remain in intersection after merging trait4
Dropped 2 SNPs due to inconsistent allele pairs from phenotype 5. 7119362 SNPs remain.
Dropped 0 SNPs due to strand ambiguity, 7119362 SNPs remain in intersection after merging trait5
... Merge of GWAS summary statistics complete. Number of SNPs:	 7119362
too many values to unpack
Traceback (most recent call last):
  File "mtag.py", line 1544, in <module>
    mtag(args)
  File "mtag.py", line 1319, in mtag
    Zs , Ns ,Fs, res_temp, DATA = extract_gwas_sumstats(DATA,args,list(np.arange(args.P)))
ValueError: too many values to unpack
Analysis terminated from error at Sun Nov 25 16:06:34 2018
Total time elapsed: 14.0m:39.51s
```

Log is attached.

[mtag.info8.out.log](https://github.com/omeed-maghzian/mtag/files/2613523/mtag.info8.out.log)

**Comment by huilisabrina** [comment]
Hi @carbocation ,

I have updated that flag to `--incld_ambig_snps` due to another separate issue https://github.com/omeed-maghzian/mtag/issues/56. Could you re-pull to the latest commit here 88c07affb53121f86e3883d27b5c4c8003d0e523. This one should resolve the args unpacking issue. Thanks! 

Best,
Hui

**Comment by carbocation** [comment]
Sweet - output has 16894 as expected now, and the table looks much more reasonable:

```
  Trait                 N (max)  N (mean)  # SNPs used  GWAS mean chi^2  MTAG mean chi^2  GWAS equiv. (max) N
1  ...lidse.gz-mtag.gz  16894    16894     7119362      1.062            1.061            16753              
2  ...lidse.gz-mtag.gz  16894    16894     7119362      1.064            1.067            17479              
3  ...-mtag.gz-100x.gz  16894    16894     7119362      1.062            1.065            17813              
4  ...lidse.gz-mtag.gz  16894    16894     7119362      1.042            1.046            18492              
5  ...lidse.gz-mtag.gz  16894    16894     7119362      1.058            1.059            17189              
```

**Comment by paturley** [comment]
@huilisabrina I think reporting the raw sample size makes a lot of sense. Thanks for updating this.

**Comment by carbocation** [comment]
Thanks again. The output now seems totally sensible, and everything runs without a hiccup.
